source/llvm/ASTNodeCodeGen.cpp

```diff
@@ -637,6 +639,159 @@ llvm::Value* ASTNodeCodeGen::delayExprCodeGen(const libsbml::ASTNode* ast)
+                        << name << " is a concentration, not an amount, but its compartment ("
+                        << compId << ") chages due to an assignment rule.  Since we cannot"
+                        << " calculate derivatives on the fly, we cannot determine the rate of"
+                        << " change of the species " << name << " concentration";
+                    throw_llvm_exception(err.str());
+                }
+                else {
+                    //The only other option is a rate rule, which means the compartment changes in time.
+                    compIsConst = false;
+                }
+            }
+        }
+
+        //The functions are different depending on how complicated things get, so we need references.  The final Value will be created from the overallRef ASTNode.
+        ASTNode amtRate(AST_PLUS);
+        ASTNode* amtRateRef = &amtRate;
+        ASTNode* overallRef = &amtRate;
+
+        //Deal with conversion factors:
+        string conversion_factor = "";
+        if (model->isSetConversionFactor()) {
+            conversion_factor = model->getConversionFactor();
+        }
+        if (species->isSetConversionFactor()) {
+            conversion_factor = species->getConversionFactor();
+        }
+        if (!conversion_factor.empty()) {
+            amtRate.setType(AST_TIMES);
+            ASTNode* cf = new ASTNode(AST_NAME);
+            ASTNode* unscaledRate = new ASTNode(AST_PLUS);
+            cf->setName(conversion_factor.c_str());
+            amtRate.addChild(unscaledRate);
+            amtRate.addChild(cf);
+            amtRateRef = unscaledRate;
+        }
+
+        //Sum all the kinetic laws * stoichiometries in which this species appears:
+        for (int rxn = 0; rxn < model->getNumReactions(); rxn++) {
+            const Reaction* reaction = model->getReaction(rxn);
+            if (reaction->getReactant(name) == NULL &&
+                reaction->getProduct(name) == NULL) {
+                continue;
+            }
+            ASTNode* times = new ASTNode(AST_TIMES);
+            ASTNode* stoich = new ASTNode(AST_NAME);
+            stoich->setName((reaction->getId() + ":" + name).c_str());
+            times->addChild(stoich);
+            times->addChild(reaction->getKineticLaw()->getMath()->deepCopy());
+            amtRateRef->addChild(times);
+        }
+
+        //Now deal with the case where the species symbol is a concentration, not an amount.
+        ASTNode scaledConcentrationRate(AST_DIVIDE);
+        ASTNode combinedConcentrationRate(AST_MINUS);
+        if (!speciesIsAmount) {
+            //Need to divide by the compartment size:
+            scaledConcentrationRate.addChild(overallRef->deepCopy());
+            ASTNode* compsize = new ASTNode(AST_NAME);
+            compsize->setName(compId.c_str());
+            scaledConcentrationRate.addChild(compsize);
+            overallRef = &scaledConcentrationRate;
```

Comment:
I would reorder these ones, too.

Reply:
I'm not sure how to reorder these; we create the compsize object, initialize it, and add it.

Reply:
I thought we could follow the same order in line 727-732 to add all the children one after another with no line in-between. But that's not a big deal, for sure.